### PR TITLE
STRING aliases and related types (CHAR, VARCHAR)

### LINKED ISSUE #7859: sql: truncate trailing spaces for char column types
PR: https://github.com/cockroachdb/cockroach/pull/50492

From release notes:
> The `CHAR` column [`type`](../v20.2/data-types.html) will now truncate the trailing space characters in line with Postgres. Existing stored `CHAR` entries with spaces at the end of the `CHAR` column [`type`](../v20.2/data-types.html) will no longer return rows with trailing space characters. Use the `LIKE` query to find and modify these rows. [#50492][#50492] {% comment %}doc{% endcomment %}

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes #7859.
Fixes https://github.com/cockroachdb/docs/issues/7622.
Fixes https://github.com/cockroachdb/docs/issues/10449.
Fixes https://github.com/cockroachdb/docs/issues/10499.

@otan ... not sure who else to add to this. 

I'm particularly skeptical about this claim I added in there:

"By default, `STRING` values are variable-length, with a limit. As such, `CHARACTER VARYING` and `VARCHAR` types are aliases for `STRING`."

Notes from 5/24/21:

- `VARCHAR` and `CHARACTER VARYING` are aliases for the same type, but have different type OIDs.
- `STRING` and `TEXT` are aliases for the same type, and have the same type OID.
- `STRING` is not in postgres. It's only in CockroachDB.
- `CHARACTER` and `CHAR` are aliases for the same type, and have the same type OID.
- `"char"` and `CHAR` are different types (like in postgres). `"char"` is limited to 1 byte of memory, and truncates strings (this currently doesn't work as it should, so wait on the bug fix).

The following are functionally identical:

- `VARCHAR`/`CHARACTER VARYING`
- `STRING`/ `TEXT`
- `CHARACTER`/`CHAR`

```
Subtype                                                               | Description                  
----------------------------------------------------------------------|------------------------------
`CHARACTER`, `CHARACTER(n)`, `CHAR`, `CHAR(n)`                        | Fixed-length                
`CHARACTER VARYING(n)`,  `VARCHAR(n)`                                 | Variable-length, with a limit  
`TEXT`, `VARCHAR`, `CHARACTER VARYING`, `STRING`                      | Variable-length, with no limit
```

See https://www.postgresql.org/docs/13/datatype-character.html.

### CODE REVIEW

**Comment by otan** [comment]
string is a CockroachDB only thing, I believe it is handled exactly like TEXT (i.e. no limit).

**Comment by ericharmeling** [comment]
Making this a WIP PR until we more carefully look at the expected differences between the subtypes

@jordanlewis

**Comment by ericharmeling** [comment]
@rafiss 

I've updated this PR following our conversation. PTAL.

File: v19.2/string.md [hunk 1]
```diff
@@ -6,22 +6,17 @@ toc: true
...
 
-- `CHARACTER`
-- `CHAR`
-- `VARCHAR`
-- `TEXT`
+Subtype                                                               | Description                  
+----------------------------------------------------------------------|------------------------------
+`CHARACTER`, `CHARACTER(n)`, `CHAR`, `CHAR(n)`                        | Fixed-length                
+`CHARACTER VARYING`, `CHARACTER VARYING(n)`, `VARCHAR`, `VARCHAR(n)`  | Variable-length, with a limit  
+`TEXT`                                                                | Variable-length, with no limit
```
Comment (rafiss) [inline v19.2/string.md:17]: i think the last two lines should look more like

```
`CHARACTER VARYING(n)`,  `VARCHAR(n)`                                 | Variable-length, with a limit  
`TEXT`, `VARCHAR`, `CHARACTER VARYING`, `STRING`                      | Variable-length, with no limit
```

(this looks correct on the docs for the other versions below)

**Review by rafiss — CHANGES_REQUESTED** [review]
just one comment, then lgtm!

File: releases/v20.2.0-alpha.2.md [hunk 2]
```diff
@@ -93,7 +93,6 @@ $ docker pull cockroachdb/cockroach-unstable:v20.2.0-alpha.2
...
 - Fixed an error for the `ALTER COLUMN TYPE` statement for situations where the data matches byte for byte but CockroachDB needs to validate the `INT8` -> `INT4` conversion. [#50278][#50278]
 - Previously, `infinity` evaluated to a negative value, `-292277022365-05-08T08:17:07Z`. Now `infinity` is the maximum supported timestamp in Postgres that is not infinity. Likewise, `-infinity` is the smallest supported value. Note that this work-in-progress feature currently does not behave exactly like `infinity` in Postgres. [#50311][#50311] {% comment %}doc{% endcomment %}
-- The `CHAR` column [`type`](../v20.2/data-types.html) will now truncate the trailing space characters in line with Postgres. Existing stored `CHAR` entries with spaces at the end of the `CHAR` column [`type`](../v20.2/data-types.html) will no longer return rows with trailing space characters. Use the `LIKE` query to find and modify these rows. [#50492][#50492] {% comment %}doc{% endcomment %}
```
Comment (ckoehn) [inline releases/v20.2.0-alpha.2.md:96]: The GA release notes of `20.2` don't mention this change at all. Shouldn't it be added there?

**Review by ericharmeling — COMMENTED** [review]
@rafiss 

TFTR !

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8250#-:-MaZX14H52az1KFXnIUS:b-mntzw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[v19.2/string.md, line 17 at r1](https://reviewable.io/reviews/cockroachdb/docs/8250#-MGj9cBh5joLQKieV8Sx-r1-17:-MaZW1msArmZh7h3AHdz:blpinxg) ([raw file](https://github.com/cockroachdb/docs/blob/f171eafcb8a635945eadefd31ac045e5afcbe589/v19.2/string.md#L17)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think the last two lines should look more like

```
`CHARACTER VARYING(n)`,  `VARCHAR(n)`                                 | Variable-length, with a limit  
`TEXT`, `VARCHAR`, `CHARACTER VARYING`, `STRING`                      | Variable-length, with no limit
```

(this looks correct on the docs for the other versions below)
</blockquote></details>

Ah. I missed the updates on this version of the page since we don't officially support 19.2 anymore.

I updated the page anyways.

Also - I added `STRING(n)` to that second row in all updated versions. Is that correct?

**Review by ericharmeling — COMMENTED** [review]
@ckoehn 

> The GA release notes of 20.2 don't mention this change at all. Shouldn't it be added there?

That's a good shout, but we try to reserve the GA release note [feature summary](https://www.cockroachlabs.com/docs/releases/v20.2.0#feature-summary) to "the most significant user-facing changes" in a GA release. And I wouldn't consider this change a significant user-facing change.

I do, however, think that this change is important enough to document in the testing release notes, and to call out on the appropriate docs page (i.e., [the `STRING` page](https://www.cockroachlabs.com/docs/v21.1/string.html)). Note that the GA release note [feature summary](https://www.cockroachlabs.com/docs/releases/v20.2.0#feature-summary) points users to testing release notes and "New in 20.2" callouts for minor user-facing updates.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8250#-:-MaZZZEw3qYB5xEYh-iM:bjelrfh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Review by rafiss — APPROVED** [review]
this lgtm! thanks for the updates

**Review by ianjevans — APPROVED** [review]

**Pull request merged**